src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -406,6 +406,242 @@ public void Dispose()
+            services.AddSingleton<Thing1>();
+            services.AddSingleton<Thing2>();
+            services.AddSingleton<Thing3>();
+            services.AddTransient(typeof(IFakeOpenGenericService<>), typeof(FakeOpenGenericService<>));
+
+            var lazy = new Lazy<Thing4>(() =>
+            {
+                thing3 = sp.GetRequiredService<Thing3>();
+                return new Thing4(thing3);
+            });
+
+            services.AddTransient(sp =>
+            {
+                if (ThreadId == 1)
+                {
+                    Thread.Sleep(1000);
+                }
+                else
+                {
+                    // Let Thread 1 over take Thread 2
+                    Thread.Sleep(3000);
+                }
+
+                return lazy.Value;
+            });
+            services.AddSingleton<Thing5>();
+
+            sp = services.BuildServiceProvider();
+
+            // Act
+            var t1 = Task.Run(() =>
+            {
+                ThreadId = 1;
+                using var scope1 = sp.CreateScope();
+                constrainedThing4Services = sp.GetServices<IFakeOpenGenericService<Thing4>>().ToList();
+            });
+
+            var t2 = Task.Run(() =>
+            {
+                ThreadId = 2;
+                using var scope2 = sp.CreateScope();
+                constrainedThing5Services = sp.GetServices<IFakeOpenGenericService<Thing5>>().ToList();
+            });
+
+            // Act
+            await t1;
+            await t2;
+
+            var thing4 = sp.GetRequiredService<Thing4>();
+            var thing5 = sp.GetRequiredService<Thing5>();
+
+            // Assert
+            Assert.NotNull(thing3);
+            Assert.NotNull(thing4);
+            Assert.NotNull(thing5);
+            Assert.Equal(1, constrainedThing4Services.Count);
+            Assert.Equal(1, constrainedThing5Services.Count);
+            Assert.Same(thing4, constrainedThing4Services[0].Value);
+            Assert.Same(thing5, constrainedThing5Services[0].Value);
+        }
+
+        private class Thing5
+        {
+            public Thing5(Thing4 thing)
+            {
+            }
+        }
+
+        private class Thing4
+        {
+            public Thing4(Thing3 thing)
+            {
+            }
+        }
+
+        private class Thing3
+        {
+            public Thing3(Thing2 thing)
+            {
+            }
+        }
+
+        private class Thing2
+        {
+            public Thing2(Thing1 thing1)
+            {
+            }
+        }
+
+        private class Thing1
+        {
+            public Thing1(Thing0 thing2)
+            {
+            }
+        }
+
+        private class Thing0
+        {
+            public Thing0()
+            {
+            }
+        }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/36458")]
+        public void GetRequiredService_CircularReference_ThrowsSOEAndHangs()
```

Comment:
I'm a bit confused by this test. We are adding a new test for an issue that we know is currently broken? Why not add this test when we fix #36458?

Reply:
sure, makes sense to remove this test.